### ISSUE TITLE
Opening a file from a folder which contains many files is slow

### ISSUE THREAD

**Issue by izuzak** [atom/atom#1847]
Halp ticket:
- support/7ca5bdc2b92d11e39039cbba6d0f0bfc

> I am opening a single python file inside a folder that contains about 29,000 files.
> ...
> I'm working on a python script to iterate through about 30,000 HTML files in a folder. So when I typed in atom compare_html.py at the command line, the atom window would open up and no file was open in the editor.
> 
> What I neglected to add to the feedback (I sent the feedback only after waiting a short while) that after two or three minutes, atom would open up the file in question and all the files would be listed in the directory tree. So it took atom a while, apparently, to load and parse the contents of the folder.
> 
> My expectation as a user would be to on open the file in the editor and not the folder. If it was going to open the folder anyway, maybe spawn that off in a different process?

I just gave this a try by creating a folder with 30000 html files (9kb each) and that does seem to be true -- Atom took a few minutes to open a file from the folder. 

File > Open completed immediately with no content displayed in the editor, but I noticed that an AtomHelper process took 100% of the CPU and starting taking more and more memory. Just after the process took 1GB of memory, a new Atom Window was shown with the contents of the file and 30k items in the tree view.

**Comment by izuzak** [comment]
Reported again in https://github.com/atom/tree-view/issues/442

**Comment by izuzak** [comment]
> Reading through here it doesn't seem like any of the previous attempts to resolve this have touched on the specific scenario I reported though.

@elneilios Yep, that's why the issue is still open. Just to make sure I understand your comment -- this _is_ the same report as the one in https://github.com/atom/tree-view/issues/442, correct?

**Comment by elneilios** [comment]
Yes I'm referring to atom/tree-view#442
I'm hoping that it can be resolved still as I don't like having to disable tree-view in order to use Atom as a simple text editor

**Comment by dkniffin** [comment]
Will this issue be addressed anytime soon? I'm having a similar issue with a rails migrations folder on my project, and it's only got 219 files.

**Comment by andrewnester** [comment]
Confirm this issue
1) When I try to expand folder that contains lots of subfolders, expanding takes about 10 seconds
2) When I try to open some file from one of these folders it looks like it takes about 5-10 seconds

I am using Atom 1.0.16
Looks like issue are related to the fact that Atom tries check all folders recursively to restore previous state of expanded folders.
It would be great if in Tree View settings we had some option to disable this behaviour

**Comment by alltom** [comment]
The tree view pops up unexpectedly when you save a new file, which led to this surprising behavior in Atom 1.0.19 on OS X 10.11:

1. Open a new Atom window. The tree view is closed.
2. Start a new file in that window.
3. Save it in a directory that contains many files.
4. Resume typing.

Expected: it's fast, the things I type go into the file I was editing, and the tree view stays closed.

Actual: the window becomes unresponsive for several seconds and my keystrokes have no effect. Then, suddenly, the tree view pops up. A different file is now open, possibly because the things I was typing during the freeze went to the tree view instead?

**Comment by keskiju** [comment]
Some plugins scan the whole project tree on startup and this is a bad thing if you just want to open a single file from your home directory. It would be nice if I could configure which plugins are disabled by default when I open just a single file.

**Comment by keskiju** [comment]
Actually there would be no problem if running "atom file" from command line would behave just like running plain "atom", which opens previous session instead of current directory. Though it might be a nice addition if atom would also provide some kind of light mode for quick single file editing which would enable only the most important plugins on startup.

**Comment by anonymousdude** [comment]
I've been experiencing this issue for ever.  About every 3 months I download Atom at work to see if the issue has been resolved yet because it makes it completely unusable for me.  Some more info... I'm using Windows 7 and the folder I'm opening is on a shared network drive that maps to a Linux server.  When I copy those same directories to my local windows machine, they expand very quickly in the tree viewer.

Unfortunately because of my company's set up, I'm stick with the shared network drive at work.

Also, this only happens on Windows.  I brought my macbook into work and connected to the same networked drive and tree viewer worked flawlessly.

**Comment by MrAdam** [comment]
I'd really like to see something done with this as well.
I used to do "subl file.ext" a lot for single file editing, and still do because "atom file.ext" is slow, and opens the tree view every time (even if I want to work on a single file).

**Comment by ScottKbka** [comment]
:+1: To fixing this. I'm new to Atom and this was quite a surprise to me. Sometimes I just want to edit one file, every directory.

**Comment by tefaa89** [comment]
This is really a deal breaker for me. I wanted to use atom for single file editing (notepad++ replacement) but I couldn't since its really really slow. Please fix.

**Comment by bryevdv** [comment]
It certainly makes it very slow to open a single file at the top of a large directory structure. In addition to that, the giant directory tree is *distracting* and *unwanted* if I just want to edit a single file.

**Comment by lexicalunit** [comment]
A setting in Tree View similar to the "Always Show Tab Bar" setting in the Tabs package would go a long way toward solving this issue for me. It would have the following behavior when unchecked:

1. If there are no Project Folders in the Tree View, don't show it.
2. If `atom file.ext` is used on the CLI to open Atom, don't automatically add the file's directory as a Project Folder.
3. If `atom /path/to/dir` is used on the CLI to open Atom, do automatically add that path as a Project Folder.
4. Similarly for the command `Application: Open File`, don't automatically add the file's directory as a Project Folder.

If the setting were checked, then we'd have the behavior we have now.

**Comment by MrAdam** [comment]
This has been an issue for over a year - any news on a fix?
I know from a lot of colleagues that it's the only thing preventing them to move completely from Sublime Text

**Comment by elneilios** [comment]
Just wanted to mention that although this problem is very annoying, I've been successfully using Atom as a file editor by completely disabling the TreeView package. I can't imagine that is very good solution for someone wanting to create entire projects with Atom however.

**Comment by lalitkapoor** [comment]
:+1: to solve this - very painful

**Comment by Kalmis** [comment]
@elneilios, if I understood the problem correctly, then there's no need to disable whole package. You can disable the tree view's auto focus by toggling Settings / Packages / Tree view / "Focus On Reveal" off. This way tree view won't automatically load folder to the view pane nor focus. You can manually focus the tree view by pressing ctrl+alt+\

**Comment by jeremyn** [comment]
A specific way to test this problem is to check out the [EFForg/https-everywhere](https://github.com/EFForg/https-everywhere) repository and then open one of the files in the [src/chrome/content/rules directory](https://github.com/EFForg/https-everywhere/tree/0822f0a76e3b56fb38288eeb20aef3cbf38d81a3/src/chrome/content/rules). (Warning: the second link might lag your browser.)

**Comment by reiven** [comment]
Any update on this??

**Comment by chrisdothtml** [comment]
Is there anything stopping atom from having an option to not load the whole tree when a file is opened? Is it because it's a massive rewrite to achieve, or just because no one's gotten around to it?

**Comment by bryevdv** [comment]
> Is it because it's a massive rewrite to achieve, or just because no one's gotten around to it?

Rather, my impression is that making the change is not wanted.

**Comment by fletom** [comment]
This is a huge usability problem, I think.

Would it not be pretty easy to add a flag to `atom` like `--single-pane` or `--no-project` to open a simple one-file editing window?

Right now, running something like `atom ~/.zshrc` is a huge pain, because it loads your entire home directory in the sidebar at the same time.

**Comment by macintacos** [comment]
Adding my voice to this. Atom is such a nice editor despite this default functionality. Only reason why I still default to Sublime

**Comment by fletom** [comment]
I would even go so far as to say that having no project tree should be the default when opening a single file, either via the GUI or via `atom`. How many times do you care about the entire directory when you're just trying to open a file in a simple text editor? `vim ~/.gitconfig` is instant. `subl ~/.gitconfig` is nearly instant. `atom ~/.gitconfig` takes about 10 seconds to finish opening and loading. That's totally unreasonable for everyday use. And what if you're trying to open a log file from a directory with 500 other log files in it?

**Comment by chrisdothtml** [comment]
@fletom Agreed. One reason I liked Sublime so much was because I could **really** quickly open up a file and edit it. Opening a single file is a huge process with Atom, and is one of the reasons I stay away from full-blown IDEs like Webstorm

**Comment by Log1x** [comment]
This is driving me absolutely crazy!

+1, please fix this!

**Comment by dbackeus** [comment]
Example of what happens when I `atom ~/.zshrc`:

<img width="685" alt="screen shot 2017-02-07 at 09 54 07" src="https://cloud.githubusercontent.com/assets/3461/22684428/9783203c-ed1c-11e6-81f4-f5dc21fd3b7f.png">

To be honest I would escalate this from a mere "performance" issue as it basically destabilizes the entire system while going on.

(by the way that beam process is spawned by the Atom "autocomplete-elixir" package and ends up being a zombie process at 160% cpu usage even after quitting Atom - why doesn't it get killed by its parent process?)

**Comment by jd20** [comment]
Same thing here, an option to disable tree view when opening single files would really help make Atom more usable for quick editing of single files (like .zshrc or .bash_profile)

**Comment by hmackiernan** [comment]
>>Is it because it's a massive rewrite to achieve, or just because no one's gotten around to it?

> Rather, my impression is that making the change is not wanted.

@bryevdv Given that people have been (quite rightly) _howling_ about this issue for **three years** I think that your impression is mistaken

**Comment by Fawers** [comment]
Yep, one more here.

Supposing a hypothetical file, `$FILE`:

* what I expect Atom to do when running `atom $FILE`:
  * open `$FILE`, **only**
* what it actually does:
  * open `$FILE`, and show EVERYTHING in `dirname $FILE` and its subdirectories on the TreeView.

I, too, come from Sublime, and even though I'm enjoying Atom a lot, this one issue could really be addressed.

**Comment by chrisdothtml** [comment]
Unfortunately I've switched to vscode because this seemingly will never be addressed. I'm not sure if atom devs have looked into how vscode handles this sort of thing, but it may be worth looking into, as it has virtually no issues in this area

**Comment by Log1x** [comment]
Daily reminder this issue makes me want to claw my eyes out.

**Comment by rgawenda** [comment]
Doing _ANYTHING_ in such a directory is a problem. It's a kernel/filesystem problem that shows in _any_ OS and near every filesystem.

**Comment by bryevdv** [comment]
> Doing ANYTHING in such a directory is a problem. It's a kernel/filesystem problem that shows in any OS and near every filesystem.

Not sure what you are trying to say. This is definitely a problem specific to Atom:

```
~ find . -name * | wc -l
27023

~ subl -n setup.py
```
SublimeText opens *instantly*, Atom would take tens of seconds or more to open that same file.

**Comment by trekky0623** [comment]
Yep, please fix this. If I'm opening a single file, I don't care about the tree. That can be reserved for opening projects.

**Comment by andremacola** [comment]
+1 for that. Moving back do Sublime

**Comment by nathggns** [comment]
Why has this still not been solved!?

**Comment by chrisdothtml** [comment]
I've been using the Atom beta for a few weeks now, and this issue has been resolved as far as I can tell. The time to open a new window is still a bit long (a few seconds), but loading a directory into Atom is pretty much instant. Glad to be able to switch back

**Comment by rusher81572** [comment]
This is also a big pain point for me. I love Atom but this is not helping my productivity.

**Comment by dougballinger** [comment]
I can't open a project folder without Atom adding in all of the files from my home folder - from apps to docs etc. Is this the same issue others are having? Very annoying.

**Comment by E-Unit** [comment]
Same issue. At work I have just over 25000 files in a network folder, most of which are less than 1Mb in size. They average at about 100-700 kb. When opening a single file in Atom, atom hangs and eventually asks me to Force close or wait. 

I just Tried disabling the tree-viewer and hen opening a file directly from the network folder and 'boom' instantly it opened. Turn tree-viewer back on with said file still open and atom freezes.

Using Atom 1.19.0 beta 5
Windows 7 pro with core i7
Windows server (if it matters)

**Comment by dougballinger** [comment]
Thanks for confirming. I've moved on to Sublime which seems to work fine. Too bad because I think Atom's UI is better but this issue renders the app unusable.

**Comment by andremacola** [comment]
=\

**Comment by dannyhanes** [comment]
+1 Would like to see this resolved.

**Comment by adrianodennanni** [comment]
I have no option other than force quit Atom when this happens. +1 to a solution.

**Comment by Log1x** [comment]
I think with over 3+ years of this going unsolved the solution is to switch to VSCode.

**Comment by frank-dspeed** [comment]
@Log1x Its not a good fix but your right the fix is valid as even i can't fix atom start up i think its a fail by design.

**Comment by asarkar** [comment]
Adding my voice in the chorus. Since so many people voting for this issue over 4 years is apparently not enough to get it fixed, the next option would be for each of us to open a new issue.

**Comment by fmdl** [comment]
HELP !!!!!!!

**Comment by E-Unit** [comment]
I've been a loyal user of Atom since its beginnings. Because this issue remains a big problem for most ppl dealing with large file numbers (not necessarily large in size) atom becomes unusable. I have recently moved on to vim and am still experiencing serious growing pains but I'm forcing myself to get good at it. Sublime it's nice but I like free stuff.
I now only use atom if I'm pressed for time.
Starting aim from cmd with --gpu-disabled and also treeview disabled seems to improve speed for those that want to stick it out

**Comment by zyend** [comment]
+1; up-voting to fix

**Comment by frank-dspeed** [comment]
I think i found the race condition that is causing this issue.

The Problem is we Would need to Refactor the Code to use Streams and manage the parallel execution this is a big big epic that needs a lot of coders.

**Comment by alebupal** [comment]
There is still no solution for this?

**Comment by frank-dspeed** [comment]
@alebupal use smaller project folders is one solution aka code splitting

**Comment by smashwilson** [comment]
With #18608 in place, opening a single file no longer implicitly opens the directory containing it as a project folder. I've tested this with opening a single file in a directory that contains 100,000 files; the single-file case is now pretty noticeably faster than opening the entire directory. (We've come a long way since this issue was opened, though 😄.)
